src/OutputPresenter.ts

```diff
@@ -0,0 +1,268 @@
+  setTrackIndex(newIndex: number): void {
+    this.trackIndex = newIndex;
+  }
+  // lets say current is long, user changes mood, and the current moodTemplate do not support long
+  getSuitableLength(filesLength: number): string {
+    if (this.canShowDuration(filesLength, this.lengthIndex)) {
+      return this.lengthIndex;
+    }
+    if (
+      this.lengthIndex === "long" &&
+      this.canShowDuration(filesLength, "medium")
+    ) {
+      return "medium";
+    }
+    return "short";
+  }
+
+  @mobx.action
+  setCurrTrack(trackIndex: number, filesLength: number): void {
+    this.trackIndex = trackIndex;
+    this.currTrack = this.currMood.tracks[trackIndex];
+    this.music.pause();
+    this.music = new Audio(this.currTrack.musicTrack);
+    this.setCurrLength(this.getSuitableLength(filesLength));
+  }
+
+  @mobx.action
+  prevCurrTrack(filesLength: number): void {
+    if (this.trackIndex - 1 < 0) {
+      this.setCurrTrack(this.currMood.tracks.length - 1, filesLength);
+    } else {
+      this.setCurrTrack(this.trackIndex - 1, filesLength);
+    }
+  }
+
+  @mobx.action
+  nextCurrTrack(filesLength: number): void {
+    if (this.trackIndex + 1 >= this.currMood.tracks.length) {
+      this.setCurrTrack(0, filesLength);
+    } else {
+      this.setCurrTrack(this.trackIndex + 1, filesLength);
+    }
+  }
+
+  @mobx.action
+  setCurrLength(newLength: string) {
+    if (
+      newLength === "short" ||
+      newLength === "medium" ||
+      newLength === "long"
+    ) {
+      this.lengthIndex = newLength;
+      this.currLength = this.currTrack[newLength];
+      this.resetVideo();
+    } else {
+      alert(newLength + " is not short, medium or long");
+    }
+  }
+
+  canShowDuration(filesLength: number, duration: string): boolean {
+    if (duration === "short") return true;
+    if (
+      duration === "medium" &&
+      this.currTrack.medium.slot.length <= filesLength * 3
+    ) {
+      return true;
+    }
+    if (
+      duration === "long" &&
+      this.currTrack.medium.slot.length <= filesLength * 3 &&
+      this.currTrack.long.slot.length <= filesLength * 3
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  @mobx.computed
+  get templateLength(): string {
+    return this.currLength.length;
+  }
+
+  resetVideo(): void {
+    this.music.load();
+    this.setMusicLoaded(false);
+    this.playedSeconds = 0;
+    this.playingMedia = 0;
+    this.overallPlayedSeconds = 0;
+    this.play = false;
+    this.seekPlayMusic();
+  }
+
+  @mobx.action
+  setMusicLoaded(loaded: boolean): void {
+    this.musicLoaded = loaded;
+  }
+
+  @mobx.action
+  toggleMusicLoaded(): void {
+    this.musicLoaded = !this.musicLoaded;
+  }
+
+  @mobx.action
+  seekPlayMusic(): void {
+    this.music.currentTime = this.currLength.start;
+    this.music.addEventListener("canplaythrough", (event) => {
+      if (this.music.readyState >= 3) {
+        this.setMusicLoaded(true);
+        this.music.volume = 0.7;
+      }
+    });
```

Comment:
are you creating an `<audio>` element? this fn should be a prop on the element, rather than being here

Reply:
the variable is `music: HTMLAudioElement = new Audio(this.currTrack.musicTrack);`, do you mean that it makes more sense to create <audio> element from a tsx file?

Reply:
Yep - this should be a React component which gets `this.currTrack.musicTrack` as its source, and then has a `canPlayThrough` fn